File: coding_skill_py/25_failsafe_pattern.py
"""Python Failsafe - Enum"""
from enum import Enum, auto

class Mode(Enum):
    NORMAL = auto()
    DEGRADED = auto()
    SAFE = auto()

class System:
    def __init__(self):
        self.mode = Mode.NORMAL
        self.errors = 0
    
    def report_error(self):
        self.errors += 1
        if self.errors >= 3 and self.mode == Mode.NORMAL:
            self.mode = Mode.DEGRADED
            print("[Failsafe] → DEGRADED")
    
    def operate(self):
        print(f"[System] Mode: {self.mode.name}")

def main():
    print("=== Python Failsafe ===")
    sys = System()
    for i in range(5):
        sys.report_error()
    sys.operate()

if __name__ == "__main__":
    main()

File: coding_skill_cpp/25_failsafe_pattern.cpp
/* C++ Failsafe - enum class */
#include <iostream>
using namespace std;

enum class Mode { NORMAL, DEGRADED, SAFE };

class System {
    Mode mode = Mode::NORMAL;
    int errors = 0;
public:
    void reportError() {
        errors++;
        if (errors >= 3 && mode == Mode::NORMAL) {
            mode = Mode::DEGRADED;
            cout << "[Failsafe] → DEGRADED" << endl;
        }
    }
    
    void operate() {
        cout << "[System] Mode: " << static_cast<int>(mode) << endl;
    }
};

int main() {
    cout << "=== C++ Failsafe ===" << endl;
    System sys;
    for (int i = 0; i < 5; ++i) sys.reportError();
    sys.operate();
    return 0;
}
